cv/lane_detection/src/lane_detection_inferencer.cpp

```diff
@@ -0,0 +1,242 @@
+    }
+
+    cv::Mat flat_image;
+    cv::hconcat(matArray, NETWORK_INPUT_CHANNEL, flat_image);
+    return flat_image;
+}
+
+Ort::SessionOptions getOrtSessionOptions(int intra_op_num_threads,
+                                         OrtCUDAProviderOptions cuda_options)
+{
+    Ort::SessionOptions session_options;
+    session_options.SetIntraOpNumThreads(intra_op_num_threads);
+    session_options.AppendExecutionProvider_CUDA(cuda_options);
+    session_options.SetGraphOptimizationLevel(
+        GraphOptimizationLevel::ORT_ENABLE_EXTENDED);
+    return session_options;
+}
+
+} // namespace
+
+LaneDetectionInferencer::LaneDetectionInferencer(std::string instance_name,
+                                                 std::string model_path)
+    : instance_name_{std::move(instance_name)}, 
+      model_path_{std::move(model_path)},
+      env_{OrtLoggingLevel::ORT_LOGGING_LEVEL_WARNING, instance_name_.c_str()},
+      cuda_options_{OrtCUDAProviderOptions()},
+      session_options_{getOrtSessionOptions(1, cuda_options_)},
+      session_{env_, model_path_.c_str(), session_options_},
+      allocator_{Ort::AllocatorWithDefaultOptions()},
+      input_names_{session_.GetInputName(0, allocator_)},
+      output_names_{session_.GetOutputName(0, allocator_)},
+      input_dims_{
+          session_.GetInputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape()},
+      output_dims_{
+          session_.GetOutputTypeInfo(0).GetTensorTypeAndShapeInfo().GetShape()},
+      input_tensor_size_(vectorProduct(input_dims_)),
+      output_tensor_size_(vectorProduct(output_dims_)),
+      output_tensor_values_{std::vector<float>(output_tensor_size_)},
+      memory_info_{Ort::MemoryInfo::CreateCpu(
+          OrtAllocatorType::OrtArenaAllocator, OrtMemType::OrtMemTypeDefault)}
+{
+}
+
+cv::Mat LaneDetectionInferencer::getLaneDetectionMask(cv::Mat input_image)
+{
+    return postprocessImage(performInference(preprocessImage(input_image)));
+}
+
+cv::Mat LaneDetectionInferencer::preprocessImage(cv::Mat input_img)
+{
+    cv::Mat resized_img;
+    cv::resize(input_img, resized_img,
+               cv::Size(NETWORK_INPUT_WIDTH, NETWORK_INPUT_HEIGHT),
+               cv::INTER_LINEAR);
+
+    cv::Mat input_gray_img;
+    cv::cvtColor(resized_img, input_gray_img, cv::COLOR_BGR2GRAY);
+    cv::Mat edges = getEdgeChannel(input_gray_img);
+    cv::Mat edges_inv = getInverseEdgeChannel(edges);
+
+    cv::Mat normalized_edges;
+    cv::Mat normalized_edges_inv;
+    edges.convertTo(normalized_edges, CV_32F, 1.0f / 255.0f, 0.0f);
+    edges_inv.convertTo(normalized_edges_inv, CV_32F, 1.0f / 255.0f, 0.0f);
+
+    cv::Mat normalized_img;
+    resized_img.convertTo(normalized_img, CV_32F, 1.0f / 255.0f, 0.0f);
+
+    return convertHWC2CHW(
+        combineImages(normalized_img, normalized_edges, normalized_edges_inv));
+}
+
+cv::Mat LaneDetectionInferencer::performInference(cv::Mat preprocessed_img)
+{
+    std::vector<float> input_tensor_values(input_tensor_size_);
```

Comment:
We can probably cache `input_tensor_values` instead of creating on each function call